File: functions/client/report.js
// todo report functions
// total nine report functions
const functions = require("firebase-functions");
const admin = require("firebase-admin");
const nearRegion = "asia-south1";

// fake profile
// only updates
exports.fakeProfile = functions
  .region(nearRegion)
  .runWith({ memory: "128MB" })
  .firestore.document(
    "Users/{userId}/Redflags/allredflags/Fakeprofile/docdata"
  )
  .onUpdate(async (change, context) => {
    try {
      const reportedMe = change.after.get("reported_me"); // number of times reported
      const uid = change.after.ref.path.split("/")[1];
      const updateTrigger = admin.firestore().doc(`Users/${uid}`); // trigger path
      if (reportedMe > 10) {
        await updateTrigger.update({
          is_disabled: false,
          is_delete: true,
        });
        await admin.auth().updateUser(uid, {disabled: false}) // revoke disable coz is_delete is tirggered
        console.log(
          `${uid} violated more than 10 triggers.So deleting the account`
        );
      } else if (reportedMe === 3) {
        await updateTrigger.update({
          is_disabled: true,
        });
        console.log(`${uid} account disabled`);
      }
    } catch (error) {
      console.log(`Error in streaming fake profile : ${error.toString()}`);
    }
  });

// inapporiatecontent/sexuallyexplicitcontent
// sexuallyexplicitcontent
// watch all CRUD
exports.sexuallyExplicitContent = functions
  .region(nearRegion)
  .runWith({ memory: "128MB" })
  .firestore.document(
    "Users/{userId}/Redflags/allredflags/Inapporiatecontent/4content/Sexuallyexplicitcontent/docdata"
  )
  .onWrite(async (change, context) => {
    try {
      const reportedMe = change.after.get("reported_me"); // number of times reported
      const uid = change.after.ref.path.split("/")[1];
      const updateTrigger = admin.firestore().doc(`Users/${uid}`); // trigger path
      if (reportedMe === 1) {
        await updateTrigger.update({
          is_delete: true,
        });
        console.log(`${uid} sending sexually violating contents`);
      } else if (reportedMe >= 2) {
        await updateTrigger.update({
          is_delete: true,
        });
        console.log(`${uid} sending sexually violating contents`);
        console.log(`Times reported : ${reportedMe}`);
      }
    } catch (error) {
      console.log(
        `Error in streaming sexuallyExplicitContent : ${error.toString()}`
      );
    }
  });

// inapporiatecontent/imagesofviolencetorture
// imagesofviolencetorture
// only updates
exports.imagesOfViolenceTorture = functions
  .region(nearRegion)
  .runWith({ memory: "128MB" })
  .firestore.document(
    "Users/{userId}/Redflags/allredflags/Inapporiatecontent/4content/Imagesofviolencetorture/docdata"
  )
  .onUpdate(async (change, context) => {
    try {
      const reportedMe = change.after.get("reported_me"); // number of times reported
      const uid = change.after.ref.path.split("/")[1];
      const updateTrigger = admin.firestore().doc(`Users/${uid}`); // trigger path
      if (reportedMe === 2) {
        await updateTrigger.update({
          is_delete: true,
        });
        console.log(`${uid} image of violence and torture`);
      } else if (reportedMe > 2) {
        await updateTrigger.update({
          is_delete: true,
        });
        console.log(`${uid} reason images of violence and torture`);
        console.log(`Times reported : ${reportedMe}`);
      }
    } catch (error) {
      console.log(
        `Error in streaming imagesOfViolenceTorture : ${error.toString()}`
      );
    }
  });

// inapporiatecontent/Hategroup
// Hategroup
// only updates
exports.hateGroup = functions
  .region(nearRegion)
  .runWith({ memory: "128MB" })
  .firestore.document(
    "Users/{userId}/Redflags/allredflags/Inapporiatecontent/4content/Hategroup/docdata"
  )
  .onUpdate(async (change, context) => {
    try {
      const reportedMe = change.after.get("reported_me"); // number of times reported
      const uid = change.after.ref.path.split("/")[1];
      const updateTrigger = admin.firestore().doc(`Users/${uid}`); // trigger path
      if (reportedMe >= 3) {
        await updateTrigger.update({
          is_delete: true,
        });
        console.log(`${uid} spreading hategroup`);
        console.log(`Times reported : ${reportedMe}`);
      }
    } catch (error) {
      console.log(`Error in streaming Hategroup : ${error.toString()}`);
    }
  });

// inapporiatecontent/Illegalactivityadvertising
// Illegalactivityadvertising
// only updates
exports.illegalActivityAdvertising = functions
  .region(nearRegion)
  .runWith({ memory: "128MB" })
  .firestore.document(
    "Users/{userId}/Redflags/allredflags/Inapporiatecontent/4content/Illegalactivityadvertising/docdata"
  )
  .onUpdate(async (change, context) => {
    try {
      const reportedMe = change.after.get("reported_me"); // number of times reported
      const uid = change.after.ref.path.split("/")[1];
      const updateTrigger = admin.firestore().doc(`Users/${uid}`); // trigger path
      if (reportedMe >= 3) {
        await updateTrigger.update({
          is_delete: true,
        });
        console.log(`${uid} spreading Illegalactivityadvertising`);
        console.log(`Times reported : ${reportedMe}`);
      }
    } catch (error) {
      console.log(
        `Error in streaming IllegalActivityAdvertising : ${error.toString()}`
      );
    }
  });

// Profileunder18
// only updates
exports.profileUnder18 = functions
  .region(nearRegion)
  .runWith({ memory: "128MB" })
  .firestore.document(
    "Users/{userId}/Redflags/allredflags/Profileunder18/docdata"
  )
  .onUpdate(async (change, context) => {
    try {
      const reportedMe = change.after.get("reported_me"); // number of times reported
      const uid = change.after.ref.path.split("/")[1];
      const updateTrigger = admin.firestore().doc(`Users/${uid}`); // trigger path
      if (reportedMe > 10) {
        await updateTrigger.update({
          is_disabled: false,
          is_delete: true,
        });
        await admin.auth().updateUser(uid, { disabled: false }); // revoke disable coz is_delete is tirggered
        console.log(
          `${uid} violated more than 10 triggers.So deleting the account`
        );
      } else if (reportedMe >= 4) {
        await updateTrigger.update({
          is_disabled: true,
        });
        console.log(`${uid} profile is under 18`);
      }
    } catch (error) {
      console.log(`Error in streaming ProfileUnder18 : ${error.toString()}`);
    }
  });

// Hatespeech
// only updates
exports.hateSpeech = functions
  .region(nearRegion)
  .runWith({ memory: "128MB" })
  .firestore.document(
    "Users/{userId}/Redflags/allredflags/Hatespeech/docdata"
  )
  .onUpdate(async (change, context) => {
    try {
      const reportedMe = change.after.get("reported_me"); // number of times reported
      const uid = change.after.ref.path.split("/")[1];
      const updateTrigger = admin.firestore().doc(`Users/${uid}`); // trigger path
      if (reportedMe > 10) {
        await updateTrigger.update({
          is_disabled: false,
          is_delete: true,
        });
        await admin.auth().updateUser(uid, { disabled: false }); // revoke disable coz is_delete is tirggered
        console.log(
          `${uid} violated more than 10 triggers.So deleting the account`
        );
      } else if (reportedMe >= 4) {
        await updateTrigger.update({
          is_disabled: true,
        });
        console.log(`${uid} spreading hate speech`);
      }
    } catch (error) {
      console.log(`Error in streaming HateSpeech : ${error.toString()}`);
    }
  });

// Fakelocation
// only updates
exports.fakeLocation = functions
  .region(nearRegion)
  .runWith({ memory: "128MB" })
  .firestore.document(
    "Users/{userId}/Redflags/allredflags/Fakelocation/docdata"
  )
  .onUpdate(async (change, context) => {
    try {
      const reportedMe = change.after.get("reported_me"); // number of times reported
      const uid = change.after.ref.path.split("/")[1];
      const updateTrigger = admin.firestore().doc(`Users/${uid}`); // trigger path
      if (reportedMe > 10) {
        await updateTrigger.update({
          is_disabled: false,
          is_delete: true,
        });
        await admin.auth().updateUser(uid, { disabled: false }); // revoke disable coz is_delete is tirggered
        console.log(
          `${uid} violated more than 10 triggers.So deleting the account`
        );
      } else if (reportedMe >= 2) {
        await updateTrigger.update({
          is_disabled: true,
        });
        console.log(`${uid} spoofing fake location`);
      }
    } catch (error) {
      console.log(`Error in streaming FakeLocation : ${error.toString()}`);
    }
  });

// Againstexploredating
// only updates
exports.againstExploreDating = functions
  .region(nearRegion)
  .runWith({ memory: "128MB" })
  .firestore.document(
    "Users/{userId}/Redflags/allredflags/Againstexploredating/docdata"
  )
  .onUpdate(async (change, context) => {
    try {
      const reportedMe = change.after.get("reported_me"); // number of times reported
      const uid = change.after.ref.path.split("/")[1];
      const updateTrigger = admin.firestore().doc(`Users/${uid}`); // trigger path

      if (reportedMe > 6) {
        await updateTrigger.update({
          is_delete: true,
        });
        console.log(`${uid} against explore dating`);
      }
    } catch (error) {
      console.log(
        `Error in streaming AgainstExploreDating : ${error.toString()}`
      );
    }
  });

File: lib/server/star_report_backend/report.dart
// @dart=2.9
// todo report backend
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:explore/server/star_report_backend/report_rtdb.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:ntp/ntp.dart';

class Report {
  static Future<void> fakeProfile(String oppositeUid) async {
    String myUid = FirebaseAuth.instance.currentUser.uid;
    try {
      ReportRTDB.reportAndBlock(oppositeUid, "Fake profile");
      DateTime nTPTimeStamp = await NTP.now();
      String fakeProfilePath =
          "Users/$oppositeUid/Redflags/allredflags/Fakeprofile/docdata";
      DocumentReference flageUser =
          FirebaseFirestore.instance.doc(fakeProfilePath);
      await flageUser.update({
        "reported_me": FieldValue.increment(1),
        "red_flag": FieldValue.arrayUnion([
          {
            "reported_uid": myUid,
            "reported_time": nTPTimeStamp,
          }
        ]),
      });
      print("Flagged $oppositeUid as fakeprofile");
    } catch (error) {
      if (error.toString().contains(
          "[cloud_firestore/not-found] Some requested document was not found.")) {
        DateTime nTPTimeStamp = await NTP.now();
        String fakeProfilePath =
            "Users/$oppositeUid/Redflags/allredflags/Fakeprofile/docdata";
        DocumentReference flageUser =
            FirebaseFirestore.instance.doc(fakeProfilePath);
        await flageUser.set({
          "reported_me": FieldValue.increment(1),
          "red_flag": FieldValue.arrayUnion([
            {"reported_uid": myUid, "reported_time": nTPTimeStamp}
          ]),
        });
        print("No red flag collection found so created one");
      } else {
        print("Error in fakeprofile : ${error.toString()}");
      }
    }
  }

  static Future<void> sexualExplicitContent(String oppositeUid) async {
    // parent of inapporiate content
     String myUid = FirebaseAuth.instance.currentUser.uid;
    try {
     
      ReportRTDB.reportAndBlock(oppositeUid, "Sexual explicit content");
      DateTime nTPTimeStamp = await NTP.now();
      String fakeProfilePath =
          "Users/$oppositeUid/Redflags/allredflags/Inapporiatecontent/4content/Sexuallyexplicitcontent/docdata";
      DocumentReference flageUser =
          FirebaseFirestore.instance.doc(fakeProfilePath);
      await flageUser.update({
        "reported_me": FieldValue.increment(1),
        "red_flag": FieldValue.arrayUnion([
          {
            "reported_uid": myUid,
            "reported_time": nTPTimeStamp,
          }
        ]),
      });
      print("Flagged $oppositeUid as fakeprofile");
    } catch (error) {
      if (error.toString().contains(
          "[cloud_firestore/not-found] Some requested document was not found.")) {
        DateTime nTPTimeStamp = await NTP.now();
        String fakeProfilePath =
            "Users/$oppositeUid/Redflags/allredflags/Inapporiatecontent/4content/Sexuallyexplicitcontent/docdata";
        DocumentReference flageUser =
            FirebaseFirestore.instance.doc(fakeProfilePath);
        await flageUser.set({
          "reported_me": FieldValue.increment(1),
          "red_flag": FieldValue.arrayUnion([
            {"reported_uid": myUid, "reported_time": nTPTimeStamp}
          ]),
        });
        print("No red flag collection found so created one");
      } else {
        print("Error in sexuallyexplicitcontent : ${error.toString()}");
      }
    }
  }

  static Future<void> imagesOfViolenceTorture(String oppositeUid) async {
    // parent of inapporiate content
     String myUid = FirebaseAuth.instance.currentUser.uid;
    try {
      ReportRTDB.reportAndBlock(oppositeUid, "Images of violence and torture");
      DateTime nTPTimeStamp = await NTP.now();
      String fakeProfilePath =
          "Users/$oppositeUid/Redflags/allredflags/Inapporiatecontent/4content/Imagesofviolencetorture/docdata";
      DocumentReference flageUser =
          FirebaseFirestore.instance.doc(fakeProfilePath);
      await flageUser.update({
        "reported_me": FieldValue.increment(1),
        "red_flag": FieldValue.arrayUnion([
          {
            "reported_uid": myUid,
            "reported_time": nTPTimeStamp,
          }
        ]),
      });
      print("Flagged $oppositeUid as fakeprofile");
    } catch (error) {
      if (error.toString().contains(
          "[cloud_firestore/not-found] Some requested document was not found.")) {
        DateTime nTPTimeStamp = await NTP.now();
        String fakeProfilePath =
            "Users/$oppositeUid/Redflags/allredflags/Inapporiatecontent/4content/Imagesofviolencetorture/docdata";
        DocumentReference flageUser =
            FirebaseFirestore.instance.doc(fakeProfilePath);
        await flageUser.set({
          "reported_me": FieldValue.increment(1),
          "red_flag": FieldValue.arrayUnion([
            {"reported_uid": myUid, "reported_time": nTPTimeStamp}
          ]),
        });
        print("No red flag collection found so created one");
      } else {
        print("Error in imagesofviolencetorture : ${error.toString()}");
      }
    }
  }

  static Future<void> hateGroup(String oppositeUid) async {
    // parent of inapporiate content
     String myUid = FirebaseAuth.instance.currentUser.uid;
    try {
      ReportRTDB.reportAndBlock(oppositeUid, "Hate group");
      DateTime nTPTimeStamp = await NTP.now();
      String fakeProfilePath =
          "Users/$oppositeUid/Redflags/allredflags/Inapporiatecontent/4content/Hategroup/docdata";
      DocumentReference flageUser =
          FirebaseFirestore.instance.doc(fakeProfilePath);
      await flageUser.update({
        "reported_me": FieldValue.increment(1),
        "red_flag": FieldValue.arrayUnion([
          {
            "reported_uid": myUid,
            "reported_time": nTPTimeStamp,
          }
        ]),
      });
      print("Flagged $oppositeUid as fakeprofile");
    } catch (error) {
      if (error.toString().contains(
          "[cloud_firestore/not-found] Some requested document was not found.")) {
        DateTime nTPTimeStamp = await NTP.now();
        String fakeProfilePath =
            "Users/$oppositeUid/Redflags/allredflags/Inapporiatecontent/4content/Hategroup/docdata";
        DocumentReference flageUser =
            FirebaseFirestore.instance.doc(fakeProfilePath);
        await flageUser.set({
          "reported_me": FieldValue.increment(1),
          "red_flag": FieldValue.arrayUnion([
            {"reported_uid": myUid, "reported_time": nTPTimeStamp}
          ]),
        });
        print("No red flag collection found so created one");
      } else {
        print("Error in hategroup : ${error.toString()}");
      }
    }
  }

  static Future<void> illegalActivityAdvertising(String oppositeUid) async {
    // parent of inapporiate content
     String myUid = FirebaseAuth.instance.currentUser.uid;
    try {
      ReportRTDB.reportAndBlock(
          oppositeUid, "Illegal activity and advertising");
      DateTime nTPTimeStamp = await NTP.now();
      String fakeProfilePath =
          "Users/$oppositeUid/Redflags/allredflags/Inapporiatecontent/4content/Illegalactivityadvertising/docdata";
      DocumentReference flageUser =
          FirebaseFirestore.instance.doc(fakeProfilePath);
      await flageUser.update({
        "reported_me": FieldValue.increment(1),
        "red_flag": FieldValue.arrayUnion([
          {
            "reported_uid": myUid,
            "reported_time": nTPTimeStamp,
          }
        ]),
      });
      print("Flagged $oppositeUid as fakeprofile");
    } catch (error) {
      if (error.toString().contains(
          "[cloud_firestore/not-found] Some requested document was not found.")) {
        DateTime nTPTimeStamp = await NTP.now();
        String fakeProfilePath =
            "Users/$oppositeUid/Redflags/allredflags/Inapporiatecontent/4content/Illegalactivityadvertising/docdata";
        DocumentReference flageUser =
            FirebaseFirestore.instance.doc(fakeProfilePath);
        await flageUser.set({
          "reported_me": FieldValue.increment(1),
          "red_flag": FieldValue.arrayUnion([
            {"reported_uid": myUid, "reported_time": nTPTimeStamp}
          ]),
        });
        print("No red flag collection found so created one");
      } else {
        print("Error in illegalactivityadvertising : ${error.toString()}");
      }
    }
  }

  static Future<void> profileUnder18(String oppositeUid) async {
     String myUid = FirebaseAuth.instance.currentUser.uid;
    try {
      ReportRTDB.reportAndBlock(oppositeUid, "Profile under 18");
      DateTime nTPTimeStamp = await NTP.now();
      String fakeProfilePath =
          "Users/$oppositeUid/Redflags/allredflags/Profileunder18/docdata";
      DocumentReference flageUser =
          FirebaseFirestore.instance.doc(fakeProfilePath);
      await flageUser.update({
        "reported_me": FieldValue.increment(1),
        "red_flag": FieldValue.arrayUnion([
          {
            "reported_uid": myUid,
            "reported_time": nTPTimeStamp,
          }
        ]),
      });
      print("Flagged $oppositeUid as fakeprofile");
    } catch (error) {
      if (error.toString().contains(
          "[cloud_firestore/not-found] Some requested document was not found.")) {
        DateTime nTPTimeStamp = await NTP.now();
        String fakeProfilePath =
            "Users/$oppositeUid/Redflags/allredflags/Profileunder18/docdata";
        DocumentReference flageUser =
            FirebaseFirestore.instance.doc(fakeProfilePath);
        await flageUser.set({
          "reported_me": FieldValue.increment(1),
          "red_flag": FieldValue.arrayUnion([
            {"reported_uid": myUid, "reported_time": nTPTimeStamp}
          ]),
        });
        print("No red flag collection found so created one");
      } else {
        print("Error in profileunder18 : ${error.toString()}");
      }
    }
  }

  static Future<void> hateSpeech(String oppositeUid) async {
     String myUid = FirebaseAuth.instance.currentUser.uid;
    try {
      ReportRTDB.reportAndBlock(oppositeUid, "Hate speech");
      DateTime nTPTimeStamp = await NTP.now();
      String fakeProfilePath =
          "Users/$oppositeUid/Redflags/allredflags/Hatespeech/docdata";
      DocumentReference flageUser =
          FirebaseFirestore.instance.doc(fakeProfilePath);
      await flageUser.update({
        "reported_me": FieldValue.increment(1),
        "red_flag": FieldValue.arrayUnion([
          {
            "reported_uid": myUid,
            "reported_time": nTPTimeStamp,
          }
        ]),
      });
      print("Flagged $oppositeUid as fakeprofile");
    } catch (error) {
      if (error.toString().contains(
          "[cloud_firestore/not-found] Some requested document was not found.")) {
        DateTime nTPTimeStamp = await NTP.now();
        String fakeProfilePath =
            "Users/$oppositeUid/Redflags/allredflags/Hatespeech/docdata";
        DocumentReference flageUser =
            FirebaseFirestore.instance.doc(fakeProfilePath);
        await flageUser.set({
          "reported_me": FieldValue.increment(1),
          "red_flag": FieldValue.arrayUnion([
            {"reported_uid": myUid, "reported_time": nTPTimeStamp}
          ]),
        });
        print("No red flag collection found so created one");
      } else {
        print("Error in hatespeech : ${error.toString()}");
      }
    }
  }

  static Future<void> fakeLocation(String oppositeUid) async {
     String myUid = FirebaseAuth.instance.currentUser.uid;
    try {
      ReportRTDB.reportAndBlock(oppositeUid, "Spoofing location");
      DateTime nTPTimeStamp = await NTP.now();
      String fakeProfilePath =
          "Users/$oppositeUid/Redflags/allredflags/Fakelocation/docdata";
      DocumentReference flageUser =
          FirebaseFirestore.instance.doc(fakeProfilePath);
      await flageUser.update({
        "reported_me": FieldValue.increment(1),
        "red_flag": FieldValue.arrayUnion([
          {
            "reported_uid": myUid,
            "reported_time": nTPTimeStamp,
          }
        ]),
      });
      print("Flagged $oppositeUid as fakeprofile");
    } catch (error) {
      if (error.toString().contains(
          "[cloud_firestore/not-found] Some requested document was not found.")) {
        DateTime nTPTimeStamp = await NTP.now();
        String fakeProfilePath =
            "Users/$oppositeUid/Redflags/allredflags/Fakelocation/docdata";
        DocumentReference flageUser =
            FirebaseFirestore.instance.doc(fakeProfilePath);
        await flageUser.set({
          "reported_me": FieldValue.increment(1),
          "red_flag": FieldValue.arrayUnion([
            {"reported_uid": myUid, "reported_time": nTPTimeStamp}
          ]),
        });
        print("No red flag collection found so created one");
      } else {
        print("Error in fakelocation : ${error.toString()}");
      }
    }
  }

  static Future<void> againstExploreDating(
      String oppositeUid, String explicitReason) async {
         String myUid = FirebaseAuth.instance.currentUser.uid;
    try {
      ReportRTDB.reportAndBlock(oppositeUid, explicitReason);
      DateTime nTPTimeStamp = await NTP.now();
      String fakeProfilePath =
          "Users/$oppositeUid/Redflags/allredflags/Againstexploredating/docdata";
      DocumentReference flageUser =
          FirebaseFirestore.instance.doc(fakeProfilePath);
      await flageUser.update({
        "reported_me": FieldValue.increment(1),
        "red_flag": FieldValue.arrayUnion([
          {
            "reported_uid": myUid,
            "reported_time": nTPTimeStamp,
          }
        ]),
      });
      print("Flagged $oppositeUid as fakeprofile");
    } catch (error) {
      if (error.toString().contains(
          "[cloud_firestore/not-found] Some requested document was not found.")) {
        DateTime nTPTimeStamp = await NTP.now();
        String fakeProfilePath =
            "Users/$oppositeUid/Redflags/allredflags/Againstexploredating/docdata";
        DocumentReference flageUser =
            FirebaseFirestore.instance.doc(fakeProfilePath);
        await flageUser.set({
          "reported_me": FieldValue.increment(1),
          "red_flag": FieldValue.arrayUnion([
            {"reported_uid": myUid, "reported_time": nTPTimeStamp}
          ]),
        });
        print("No red flag collection found so created one");
      } else {
        print("Error in againstexploredating : ${error.toString()}");
      }
    }
  }
}
